File: src/github_batch_manager/store.py
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Iterable

from github_batch_manager.models import RepositoryRecord


def default_store_path() -> Path:
    appdata_dir = os.getenv("APPDATA")
    if appdata_dir:
        return Path(appdata_dir) / "GitHubBatchManager" / "repositories.json"
    return Path.home() / ".github-batch-manager" / "repositories.json"


class RepositoryStore:
    def __init__(self, file_path: Path | None = None) -> None:
        self.file_path = file_path or default_store_path()

    def load(self) -> list[RepositoryRecord]:
        repositories, _ui_state = self.load_app_state()
        return repositories

    def load_app_state(self) -> tuple[list[RepositoryRecord], dict[str, object]]:
        if not self.file_path.exists():
            return [], {}

        try:
            payload = json.loads(self.file_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return [], {}

        repositories: list[RepositoryRecord] = []
        ui_state: dict[str, object] = {}

        if isinstance(payload, dict):
            repo_items = payload.get("repositories", [])
            raw_ui_state = payload.get("ui_state", {})
            if isinstance(raw_ui_state, dict):
                ui_state = raw_ui_state
        else:
            repo_items = payload if isinstance(payload, list) else []

        for item in repo_items if isinstance(repo_items, list) else []:
            if not isinstance(item, dict) or "path" not in item:
                continue
            repositories.append(RepositoryRecord.from_store_dict(item))

        return repositories, ui_state

    def save(self, repositories: Iterable[RepositoryRecord]) -> None:
        self.save_app_state(repositories, {})

    def save_app_state(
        self,
        repositories: Iterable[RepositoryRecord],
        ui_state: dict[str, object],
    ) -> None:
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        temp_path = self.file_path.with_suffix(".tmp")
        serialized = {
            "repositories": [repository.to_store_dict() for repository in repositories],
            "ui_state": ui_state,
        }
        temp_path.write_text(
            json.dumps(serialized, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )
        temp_path.replace(self.file_path)

File: src-tauri/src/store.rs
use crate::models::{AppState, AppUiState, RepositoryRecord, StoredRepositoryRecord};
use serde_json::Value;
use std::fs;
use std::io;
use std::path::PathBuf;

pub struct RepositoryStore {
    file_path: PathBuf,
}

impl RepositoryStore {
    pub fn new() -> Self {
        Self {
            file_path: default_store_path(),
        }
    }

    pub fn load(&self) -> AppState {
        let Ok(raw) = fs::read_to_string(&self.file_path) else {
            return AppState {
                repositories: Vec::new(),
                ui_state: AppUiState::default(),
            };
        };

        let Ok(payload) = serde_json::from_str::<Value>(&raw) else {
            let _ = self.backup_invalid_store("invalid", &raw);
            return AppState {
                repositories: Vec::new(),
                ui_state: AppUiState::default(),
            };
        };

        let repo_items = if let Some(items) = payload.get("repositories") {
            items.clone()
        } else if payload.is_array() {
            payload.clone()
        } else {
            Value::Array(Vec::new())
        };

        let repositories = repo_items
            .as_array()
            .map(|items| {
                items
                    .iter()
                    .filter_map(|item| {
                        serde_json::from_value::<StoredRepositoryRecord>(item.clone()).ok()
                    })
                    .map(RepositoryRecord::from)
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default();

        let ui_state = payload
            .get("ui_state")
            .cloned()
            .and_then(|item| serde_json::from_value::<AppUiState>(item).ok())
            .unwrap_or_default();

        AppState {
            repositories,
            ui_state,
        }
    }

    pub fn save(
        &self,
        repositories: &[RepositoryRecord],
        ui_state: &AppUiState,
    ) -> Result<(), io::Error> {
        if let Some(parent) = self.file_path.parent() {
            fs::create_dir_all(parent)?;
        }

        let temp_path = self.file_path.with_extension("tmp");
        let stored_repositories = repositories
            .iter()
            .map(StoredRepositoryRecord::from)
            .collect::<Vec<_>>();
        let serialized = serde_json::json!({
            "repositories": stored_repositories,
            "ui_state": ui_state,
        });

        fs::write(
            &temp_path,
            serde_json::to_string_pretty(&serialized).expect("store payload should serialize"),
        )?;
        if self.file_path.exists() {
            let backup_path = self.file_path.with_extension("bak");
            let _ = fs::remove_file(&backup_path);
            fs::rename(&self.file_path, &backup_path)?;
            match fs::rename(&temp_path, &self.file_path) {
                Ok(()) => {
                    let _ = fs::remove_file(backup_path);
                }
                Err(error) => {
                    let _ = fs::rename(&backup_path, &self.file_path);
                    let _ = fs::remove_file(&temp_path);
                    return Err(error);
                }
            }
        } else {
            fs::rename(temp_path, &self.file_path)?;
        }
        Ok(())
    }

    fn backup_invalid_store(&self, suffix: &str, raw: &str) -> io::Result<()> {
        if let Some(parent) = self.file_path.parent() {
            fs::create_dir_all(parent)?;
        }
        let backup_path = self.file_path.with_extension(format!("{suffix}.json"));
        fs::write(backup_path, raw)
    }
}

fn default_store_path() -> PathBuf {
    if let Some(appdata) = std::env::var_os("APPDATA") {
        return PathBuf::from(appdata)
            .join("GitHubBatchManager")
            .join("repositories.json");
    }

    home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".github-batch-manager")
        .join("repositories.json")
}

fn home_dir() -> Option<PathBuf> {
    std::env::var_os("USERPROFILE")
        .or_else(|| std::env::var_os("HOME"))
        .map(PathBuf::from)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::RepositoryRecord;

    fn unique_temp_dir(label: &str) -> PathBuf {
        std::env::temp_dir().join(format!(
            "github-batch-manager-{label}-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .expect("system clock before unix epoch")
                .as_nanos()
        ))
    }

    struct TempDirGuard(PathBuf);

    impl Drop for TempDirGuard {
        fn drop(&mut self) {
            let _ = std::fs::remove_dir_all(&self.0);
        }
    }

    #[test]
    fn load_preserves_old_file_visibility_for_invalid_json() {
        let root = unique_temp_dir("store");
        std::fs::create_dir_all(&root).expect("create temp dir");
        let _guard = TempDirGuard(root.clone());
        let store_path = root.join("repositories.json");

        std::fs::write(&store_path, "{invalid json").expect("write invalid store");

        let store = RepositoryStore {
            file_path: store_path,
        };

        let state = store.load();

        assert!(state.repositories.is_empty());
        assert_eq!(state.ui_state, AppUiState::default());
    }

    #[test]
    fn load_keeps_backup_of_invalid_store_file() {
        let root = unique_temp_dir("store-invalid");
        std::fs::create_dir_all(&root).expect("create temp dir");
        let _guard = TempDirGuard(root.clone());
        let store_path = root.join("repositories.json");

        std::fs::write(&store_path, "{invalid json").expect("write invalid store");

        let store = RepositoryStore {
            file_path: store_path.clone(),
        };

        let _state = store.load();

        let backup_path = root.join("repositories.invalid.json");
        assert!(backup_path.exists(), "invalid payload should be backed up");
    }

    #[test]
    fn save_writes_recoverable_store_payload() {
        let root = unique_temp_dir("store-save");
        std::fs::create_dir_all(&root).expect("create temp dir");
        let _guard = TempDirGuard(root.clone());
        let store_path = root.join("repositories.json");
        let store = RepositoryStore {
            file_path: store_path.clone(),
        };

        store
            .save(&[RepositoryRecord::new("C:/repo")], &AppUiState::default())
            .expect("save state");

        let raw = std::fs::read_to_string(store_path).expect("read store");
        assert!(raw.contains("\"repositories\""));
        assert!(raw.contains("\"uiState\"") || raw.contains("\"ui_state\""));
    }
}
